src/core/server/integration_tests/saved_objects/migrations/group3/read_batch_size.test.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Path from 'path';
+import fs from 'fs/promises';
+import { Root } from '@kbn/core-root-server-internal';
+import {
+  createRootWithCorePlugins,
+  createTestServers,
+  type TestElasticsearchUtils,
+} from '@kbn/core-test-helpers-kbn-server';
+import { delay } from '../test_utils';
+
+const logFilePath = Path.join(__dirname, 'read_batch_size.log');
+
+describe('migration v2 - read batch size', () => {
+  let esServer: TestElasticsearchUtils;
+  let root: Root;
+  let logs: string;
+
+  beforeEach(async () => {
+    await fs.unlink(logFilePath).catch(() => {});
+  });
+
+  afterEach(async () => {
+    await root?.shutdown();
+    await esServer?.stop();
+    await delay(10);
+  });
+
+  it('reduces the read batchSize in half if a batch exceeds maxReadBatchSizeBytes', async () => {
+    const { startES } = createTestServers({
+      adjustTimeout: (t: number) => jest.setTimeout(t),
+      settings: {
+        es: {
+          license: 'basic',
+          dataArchive: Path.join(__dirname, '..', 'archives', '8.4.0_with_sample_data_logs.zip'),
+        },
+      },
+    });
+
+    root = createRoot({ maxReadBatchSizeBytes: 15000 });
+    esServer = await startES();
+    await root.preboot();
+    await root.setup();
+    await root.start();
+
+    // Check for migration steps present in the logs
+    logs = await fs.readFile(logFilePath, 'utf-8');
+
+    expect(logs).toMatch(
+      'Read a batch that exceeds migrations.maxReadBatchSizeBytes, retrying by reducing the batch size in half to 15'
```

Comment:
I may have missed it, but do we have a test for recovering back to default size?

Reply:
We have https://github.com/elastic/kibana/blob/migrations-dynamic-read-batchsize/packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.test.ts#L2424 that tests that it increases from 500 to 600 on success. I can expand that to test that it increases up to 1000.